RSDK-3757 Use hardware encoder on a RPi

bazile-clyde opened the pull request:
# Description
This PR allows the RDK to use the h.264 hardware encoder on Raspberry Pis (see [RSDK-3205](https://viam.atlassian.net/browse/RSDK-3205)) by including a static, stripped-down build of FFmpeg. 

There is no need to review the source and archive files for FFmpeg. Nothing was changed.

# Testing
On all platforms, excluding the Raspberry Pi, nothing changes. For the Raspberry Pi, it now uses the h.264 hardware encoder by default. To test this, start the viam-server with a camera component and view the live stream since the live stream is the only thing that uses the h.264 MIME type and, subsequently, the hardware encoder on RPis. You should see output like the following in the logs.

```
2023-11-02T20:34:40.252-0400    INFO    gostream/stream.go:261  detected new image bounds       {"width": 640, "height": 480}
[h264_v4l2m2m @ 0x7f44000b60] Using device /dev/video11
[h264_v4l2m2m @ 0x7f44000b60] driver 'bcm2835-codec' on card 'bcm2835-codec-encode' in mplane mode
[h264_v4l2m2m @ 0x7f44000b60] requesting formats: output=YU12 capture=H264
[h264_v4l2m2m @ 0x7f44000b60] Failed to set timeperframe
```
On all other platforms, you should still have the x264 software encoder library in use.

```
2023-11-02T20:47:38.404-0400    INFO    gostream/stream.go:261  detected new image bounds       {"width": 640, "height": 480}
x264 [info]: using cpu capabilities: MMX2 SSE2Fast SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2
x264 [info]: profile Constrained Baseline, level 3.0, 4:2:0, 8-bit
```
## Platforms
I tested this code on the following 64-bit platforms.
| Name  | OS | hasHardwareEncoder | Architecture |
| ------------- | -- | ------------- | --- |
| Framework Laptop  | 22.04.3 LTS (Jammy Jellyfish) | :x: | x86_64
| Jetson Orin Nano | 20.04.6 LTS (Focal Fossa)  | :x: | aarch64
| Raspberry Pi | Debian GNU/Linux 11 (bullseye)  | :white_check_mark: | aarch64


## Storage
This change increases the size of the AppImage by about 460kb for arm64 builds. I built the AppImages using the command  `canon -arch arm64 make server-static` on each branch, tested them both to ensure they performed as expected, and below are the results.
```console
pi@pi:~ $ ls -l
total 158224
-rwxr-xr-x 1 pi pi 80775448 Nov  6 22:09 viam-server.main
-rwxr-xr-x 1 pi pi 81236056 Nov  6 21:52 viam-server.RSDK-3757
pi@pi:~ $ ls -lh
total 155M
-rwxr-xr-x 1 pi pi 78M Nov  6 22:09 viam-server.main
-rwxr-xr-x 1 pi pi 78M Nov  6 21:52 viam-server.RSDK-3757
```

# Non-goals
* It may be the case the other platforms have hardware encoders, and if so, we'll want to expand this functionality to those platforms. That's out of the scope of this project.
* There are lots of options to improve the performance of the encoder. For now, I'm choosing sensible defaults. Later, we may want to consider changing some of the hard-coded values like the bitrate and b-frame rate.

[RSDK-3205]: https://viam.atlassian.net/browse/RSDK-3205?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

bazile-clyde commented on components/camera/platforms/jetson/camera.go:
```diff
@@ -1,66 +1,18 @@
 package jetsoncamera
 
-// #include <sys/utsname.h>
```
@seanavery This was really helpful so I moved it to the utils folder to generalize it. No functionality was changed.

bazile-clyde commented on web/server/entrypoint_arm64.go:
```diff
@@ -0,0 +1,27 @@
+	"strings"
+)
+
+var streamConfig gostream.StreamConfig
```
Nothing outside this file uses this variable, so we don't need a lock.

bazile-clyde replied:
So we only ever call it once. It reads and parses a file, so I wouldn't want to add that overhead to `makeStreamConfig`.

Otterverse replied:
Right, but makeStreamConfig is only ever called once anyway AFAIK, so I don't think there's any savings there. And keeping it in a function (like every other entrypoint does) saves keeping a variable on the stack for the entire lifetime of the process. Granted, that's tiny and trivial, but I do think it's cleaner not to have init() "magic" where it can be avoided, and a package-level variable means other code anywhere in this package that might use the variable name "streamConfig" is easier to break. E.g. by accidentally using this variable when they shouldn't, because auto-complete in an IDE just filled it in.

bazile-clyde replied:
Good point. This has been updated.

seanavery commented on web/server/entrypoint_arm64.go:
```diff
@@ -0,0 +1,27 @@
+var streamConfig gostream.StreamConfig
+
+func init() {
+	if osInfo, err := utils.DetectOSInformation(); err == nil && strings.Contains(osInfo.Device, "Raspberry Pi") {
```
Can this not be added to `entrypoint_unix.go`?

bazile-clyde replied:
This line _will_ build on any Unix architecture. The line below it will _not_. Since the `h264` package depends on static libraries built on the arm64, attempting to build it on other platforms will result in arm64 symbols being undefined and the build will fail.

seanavery commented on gostream/codec/h264/ffmpeg/avcodec/avcodec.go:
```diff
@@ -0,0 +1,272 @@
+func (ctxt *Context) SetEncodeParams(width, height int, pxlFmt PixelFormat, hasBFrames bool, gopSize int) {
+	ctxt.width = C.int(width)
+	ctxt.height = C.int(height)
+	ctxt.bit_rate = 200000
```
Is this 200 Kbps? Seems reasonable for small image sizes but wondering if it is enough for 1080p+ frame sizes. How is the video quality for larger frames? Maybe we can scale bit_rate by size if necessary.

bazile-clyde replied:
This is a good point. I'll create a follow up tickets to see how these parameters impact performance but 10x this gave us as clear of an image we need right now for large images. For context, I'm using an E-Meet with a res of 1920x180.

seanavery commented on gostream/codec/h264/encoder.go:
```diff
@@ -0,0 +1,145 @@
+		return nil, errors.Wrap(err, "cannot read image")
+	}
+
+	h.frame.SetFrameFromImg(yuvImg.(*image.YCbCr))
```
Nice! glad we are able to pass in YCbCr here.

seanavery commented on gostream/codec/h264/encoder.go:
```diff
@@ -0,0 +1,145 @@
+		return nil, errors.New("cannot allocate video codec context")
+	}
+
+	h.context.SetEncodeParams(width, height, avcodec.PixelFormat(pixelFormat), false, 10)
```
Are we not able to send `bitrate` as param here?

seanavery replied:
I see that optimizing params is not a goal of this PR.

seanavery reviewed (commented):
Awesome, this LGTM.

Just added a couple of questions for my understanding.

bazile-clyde commented on web/server/entrypoint_arm64.go:
```diff
@@ -0,0 +1,27 @@
+//go:build cgo
```
Because it's in the filename. `_arm64.go` ensures it only builds for that arch.

bazile-clyde replied:
Added build tags to leave the android pathway as is. I used the `android` build tag rather than `no_cgo` since it's clearer, and both appear to only be used for `server-android` target. Let me know if there's a reason I should use the `no_cgo` tag instead.

Otterverse commented on gostream/codec/h264/ffmpeg/avutil/avutil.go:
```diff
@@ -0,0 +1,102 @@
+package avutil
+
+//#cgo CFLAGS: -Wno-deprecated-declarations -I${SRCDIR}/../ffmpeg/include
+//#cgo LDFLAGS: -L${SRCDIR}/../ffmpeg/lib -lavformat -lavcodec -lavutil -lavdevice -lavfilter -lswresample -lswscale -lm -lX11
```
```suggestion
//#cgo LDFLAGS: -L${SRCDIR}/../ffmpeg/lib -lavutil -lswresample
```
I don't think you need all those other libraries.

bazile-clyde replied:
Done.

Otterverse commented on gostream/codec/h264/ffmpeg/avcodec/avcodec.go:
```diff
@@ -0,0 +1,272 @@
+package avcodec
+
+//#cgo CFLAGS: -Wno-deprecated-declarations -I${SRCDIR}/../ffmpeg/include
+//#cgo LDFLAGS: -L${SRCDIR}/../ffmpeg/lib -lavformat -lavcodec -lavutil -lavdevice -lavfilter -lswresample -lswscale -lm -lX11
```
```suggestion
//#cgo LDFLAGS: -L${SRCDIR}/../ffmpeg/lib -lavformat -lavcodec -lavutil
```
As with avutil.go I think you're including a lot of unused libraries.

bazile-clyde replied:
Done.

Otterverse commented on gostream/codec/h264/ffmpeg/ffmpeg/share/ffmpeg/examples/Makefile:
Do we need to include all these examples in our repo?

bazile-clyde replied:
Nope. I tested this some more and got rid of everything I could. I also added some more lines to the `.gitignore` file in the `ffmpeg` directory so we don't accidentally add them back.

Otterverse commented on gostream/codec/h264/ffmpeg/ffmpeg/lib/libavcodec.a:
[optional] As these are all specifically arm64, it might be good to rename this folder lib_arm64 or lib/arm64 as a subfolder.

bazile-clyde replied:
Done! I added a command in the Makefile that does the building and ignores unnecessary files. `make ffmpeg-h264-static` will add everything to `gostream/codec/h264/ffmpeg/$(shell uname -s)-$(shell uname -m)`

Otterverse reviewed (changes requested):
A bunch of notes, mostly around build process stuff and detection logic. I'm leaving the codec specific stuff up to others on the AV team to nitpick.

I mention a couple different suggestions for the detection method, but I think just looking for the /sys/bus... file is probably the cleanest, and then no need to move the DetectOS function out of the jetson stuff (as it's really set up in a way that's linux/arm64 and even jetson specific) and as you grow the codec to support other hardware acceleration stuff, you'll likely need to customize your own detection code in a potentially different direction anyway.

One other request as you move forward is to make sure to test `make server-static` as well as the normal appimage build path, and ensure both work on a "clean" pi. It all looks good currently in my testing, but static builds are where CGO stuff can sometimes get wonky.

bazile-clyde commented:
> A bunch of notes, mostly around build process stuff and detection logic. I'm leaving the codec specific stuff up to others on the AV team to nitpick.
> 
> I mention a couple different suggestions for the detection method, but I think just looking for the /sys/bus... file is probably the cleanest, and then no need to move the DetectOS function out of the jetson stuff (as it's really set up in a way that's linux/arm64 and even jetson specific) and as you grow the codec to support other hardware acceleration stuff, you'll likely need to customize your own detection code in a potentially different direction anyway.
> 
> One other request as you move forward is to make sure to test `make server-static` as well as the normal appimage build path, and ensure both work on a "clean" pi. It all looks good currently in my testing, but static builds are where CGO stuff can sometimes get wonky.

Thanks for the notes...

I moved the OS stuff back to the Jetson directory and instead of checking for "Raspberry Pi" I now check for the existence of the hardware encoder we support (h264_v4l2m2m) on the architectures we support (arm64) which eliminated the need to check OS information (besides the build tags) entirely.

I also tested this on a _fresh_ Raspberry Pi install (bookworm) and used `server-static` this time instead of just `appimage`. I updated the PR comment to reflect that. It has size changes too. This PR adds about 460kb.

bazile-clyde commented on gostream/codec/h264/ffmpeg/ffmpeg/lib/pkgconfig/libavcodec.pc:
Nope. Removed.

bazile-clyde commented on web/server/entrypoint_arm64.go:
```diff
@@ -0,0 +1,28 @@
+var streamConfig gostream.StreamConfig
+
+func init() {
+	if osInfo, err := utils.DetectOSInformation(); err == nil && strings.Contains(osInfo.Device, "Raspberry Pi") {
```
Done. Opted to explicitly check for what we need, the codec itself.

seanavery commented on gostream/codec/h264/encoder.go:
```diff
@@ -0,0 +1,148 @@
+			bytes = append(bytes, payload...)
+			pkt.Unref()
+		case avutil.ErrorEAGAIN:
+			break loop
```
One question here on error handling. 

Given that this loop appears to accumulate the packets for a given encoded frame. Is it possible that a one-off `EAGAIN` occurs from `ReceivePacket` but the next one is successful? In such a case, I am wondering if it makes more sense to continue here.

bazile-clyde replied:
Good question. Here, I'm just doing exactly what the docs are telling me to do. From [here](https://ffmpeg.org/doxygen/3.4/group__lavc__encdec.html)
> For encoding, call [avcodec_receive_packet()](https://ffmpeg.org/doxygen/3.4/group__lavc__decoding.html#ga5b8eff59cf259747cf0b31563e38ded6). On success, it will return an [AVPacket](https://ffmpeg.org/doxygen/3.4/structAVPacket.html) with a compressed frame. Repeat this call until it returns [AVERROR(EAGAIN)](https://ffmpeg.org/doxygen/3.4/group__lavu__error.html#gae4bb6f165973d09584e0ec0f335f69ca) or an error. The [AVERROR(EAGAIN)](https://ffmpeg.org/doxygen/3.4/group__lavu__error.html#gae4bb6f165973d09584e0ec0f335f69ca) return value means that new input data is required to return new output.

If something goes wrong here it shouldn't throw an EAGAIN error but another error. That other error we'll return.

bazile-clyde replied:
I added that link as a comment to this loop.

seanavery commented on gostream/codec/h264/encoder.go:
```diff
@@ -0,0 +1,148 @@
+		case avutil.ErrorEAGAIN:
+			break loop
+		default:
+			return nil, avutil.ErrorFromCode(ret)
```
I am wondering here whether there are other error cases where we would want to return the accumulated bytes here instead of returning nil. Curious if `avutil.ErrorEOF` ever gets hit.

bazile-clyde replied:
Yes! Great catch. An EOF may be thrown if we're in draining mode - see the link in the other comment if you're curious - which is enabled by sending a `nil` frame. Then, we'll loop until we get an EOF error, not an EAGAIN, which drains all the buffered frames. We don't explicitly enable draining mode, but better to be safe here. 

There were no more errors mentioned in the docs, so I only added the EOF.

bazile-clyde replied:
Done!

Otterverse reviewed (approved):
LGTM, thanks for all the cleanup on things!
